src/io/video.js

```diff
@@ -0,0 +1,332 @@
+const log = require('../util/log');
+
+class Video {
+    constructor (runtime) {
+        /**
+         * Reference to the owning Runtime.
+         * @type{!Runtime}
+         */
+        this.runtime = runtime;
+
+        /**
+         * Cache frames for this many ms.
+         * @type number
+         */
+        this._frameCacheTimeout = 16;
+
+        /**
+         * Store each request for video, so when all are released we can disable preview/video feed.
+         * @type Array.<object>
+         */
+        this._requests = [];
+
+        /**
+         * DOM Video element
+         * @private
+         */
+        this._video = null;
+
+        /**
+         * Usermedia stream track
+         * @private
+         */
+        this._track = null;
+
+        /**
+         * Stores some canvas/frame data per resolution/mirror states
+         */
+        this._workspace = [];
+
+        /**
+         * Id representing a Scratch Renderer skin the video is rendered to for
+         * previewing.
+         * @type {number}
+         */
+        this._skinId = -1;
+
+        /**
+         * The Scratch Renderer Skin object.
+         * @type {Skin}
+         */
+        this._skin = null;
+
+        /**
+         * Id for a drawable using the video's skin that will render as a video
+         * preview.
+         * @type {Drawable}
+         */
+        this._drawable = -1;
+    }
+
+    static get FORMAT_IMAGE_DATA () {
+        return 'image-data';
+    }
+
+    static get FORMAT_CANVAS () {
+        return 'canvas';
+    }
+
+    /**
+     * Dimensions the video stream is analyzed at after its rendered to the
+     * sample canvas.
+     * @type {Array.<number>}
+     */
+    static get DIMENSIONS () {
+        return [480, 360];
+    }
+
+    /**
+     * Order preview drawable is inserted at in the renderer.
+     * @type {number}
+     */
+    static get ORDER () {
+        return 1;
+    }
+
+    /**
+     * Request video be enabled.  Sets up video, creates video skin and enables preview.
+     *
+     * ioDevices.video.requestVideo()
+     *   .then(({ release }) => {
+     *     this.releaseVideo = release;
+     *   })
+     *
+     * @return {Promise.<VideoRequest>} A request object with a "release" property that
+     *      should be called when you are done with the video.
+     */
+    requestVideo () {
+        const io = this;
+        const request = {
+            release () {
+                const index = io._requests.indexOf(request);
+                if (index > -1) {
+                    io._requests.splice(index, 1);
+                }
+                if (io._requests.length === 0) {
+                    io._disablePreview();
+                    // by clearing refs to video and track, we should lose our hold over the camera
+                    io._video = null;
+                    io._track = null;
+                }
+            }
+        };
+
+        if (this.videoReady) {
```

Comment:
I smell a race condition here. `this._lastSetup` can resolve and the unsetting callback execute before `videoReady` is `true`. Instead of unsetting `this._lastSetup` keep it and `.then` off it instead of `Promise.resolve`. This will decrease the code surface in the method too.

Reply:
I think with the refactor you also suggested, this comment is less important... `_singleSetup` only unsets and allows trying a fresh request if a previous one errors.